eZ/Bundle/EzPublishCoreBundle/DependencyInjection/Configuration.php

```diff
@@ -73,25 +73,93 @@ public function addRepositoriesSection( ArrayNodeDefinition $rootNode )
                     ->useAttributeAsKey( 'alias' )
                     ->prototype( 'array' )
                         ->beforeNormalization()
-                            // If set to null, use default values.
-                            // %ezpublish.api.storage_engine.default% as engine, and default connection (if applicable).
-                            ->ifNull()
-                            ->then(
-                                function ()
+                            ->always(
+                                function ( $v )
                                 {
-                                    return array( 'engine' => '%ezpublish.api.storage_engine.default%', 'connection' => null );
+                                    if ( isset( $v['storage'] ) )
+                                    {
+                                        return $v;
+                                    }
+
+                                    if ( isset( $v['engine'] ) )
+                                    {
+                                        $v['storage']['engine'] = $v['engine'];
+                                        unset( $v['engine'] );
+                                    }
+
+                                    if ( isset( $v['connection'] ) )
+                                    {
+                                        $v['storage']['connection'] = $v['connection'];
+                                        unset( $v['connection'] );
+                                    }
+
+                                    if ( isset( $v['config'] ) )
+                                    {
+                                        $v['storage']['config'] = $v['config'];
+                                        unset( $v['config'] );
+                                    }
+
+                                    return $v;
+                                }
+                            )
+                        ->end()
+                        ->beforeNormalization()
```

Comment:
Because I can :) First one takes care of translating old config structure, second one sets default values, as they can't be set below on concrete array nodes.